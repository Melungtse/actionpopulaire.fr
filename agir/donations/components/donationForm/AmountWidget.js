import React from "react";
import PropTypes from "prop-types";

import AmountInput from "@agir/donations/donationForm/AmountInput";
import { displayPrice } from "@agir/lib/utils/display";
import styled from "styled-components";

import "./style.css";
import { FlexContainer } from "./elements";

const DEFAULT_AMOUNTS = [200 * 100, 100 * 100, 50 * 100, 20 * 100, 10 * 100];

const AmountButton = styled.button`
  display: block;
  width: 30%;
  height: 50px;
  margin: 10px 0;
  font-weight: bold;
  min-width: 200px;
`;

class AmountWidget extends React.Component {
  constructor(props) {
    super();

    const amountChoices = props.amountChoices || DEFAULT_AMOUNTS;

    const custom = props.amount && !amountChoices.includes(props.amount);

    this.state = {
      custom,
    };
  }

  updateWithButton(value) {
    this.setState({ custom: false });
    this.props.onAmountChange(value);
  }

  updateWithCustomValue(value) {
    this.setState({ custom: true });
    this.props.onAmountChange(value);
  }

  render() {
    const { custom } = this.state;
    const { disabled, amount, error, showTaxCredit, byMonth } = this.props;

    const amountChoices = this.props.amountChoices || DEFAULT_AMOUNTS;

    return (
      <div className="amount-component">
        <FlexContainer className={error ? " has-error" : ""}>
          {amountChoices.map((value) => (
            <AmountButton
              disabled={disabled}
              key={value}
              type="button"
              onClick={() => this.updateWithButton(value)}
              className={[
                "btn",
                custom || amount !== value ? "btn-default" : "btn-primary",
              ].join(" ")}
            >
              {displayPrice(value)}
            </AmountButton>
          ))}
          <AmountInput
            disabled={disabled}
            placeholder="autre montant"
            onChange={this.updateWithCustomValue.bind(this)}
            value={custom ? amount : null}
          />

          {error && <span className="help-block">{error}</span>}
        </FlexContainer>
        <p>
          {showTaxCredit &&
            (amount ? (
              <em>
                Si je paye des impôts, après réduction, ma contribution nette
                sera de seulement{" "}
                <strong className="text-danger">
                  {displayPrice(amount * 0.34)}
                </strong>
                {byMonth && " par mois"}
                &nbsp;!
              </em>
            ) : (
              <em>
                Si je paye des impôts, je bénéficie d&apos;une réduction
                d&apos;impôt de <strong>66&nbsp;%</strong> de la somme donnée !
              </em>
            ))}
        </p>
      </div>
    );
  }
}

AmountWidget.propTypes = {
  disabled: PropTypes.bool,
  amount: PropTypes.number,
  onAmountChange: PropTypes.func,
  error: PropTypes.string,
  amountChoices: PropTypes.arrayOf(PropTypes.number),
  showTaxCredit: PropTypes.bool,
  byMonth: PropTypes.bool,
};

export default AmountWidget;
