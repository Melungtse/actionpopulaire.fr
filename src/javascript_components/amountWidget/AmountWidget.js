import React from 'react';
import {hot} from 'react-hot-loader';
import PropTypes from 'prop-types';
import {InputGroup} from 'react-bootstrap';


import './style.css';


const AMOUNTS = [100, 50, 25, 15, 10];

function displayNumber(n) {
  const s = Math.round(n * 100).toString();
  return (s.slice(0, -2) | '0') + ',' + s.slice(-2);
}

function toFloat(n) {
  return parseFloat(n.replace(',', '.'));
}


class AmountWidget extends React.Component {
  constructor({hiddenField}) {
    super();

    this.hiddenField = hiddenField;

    this.state = {
      value: null,
      custom: false
    };
  }

  updateWithCustomValue(s) {
    if (s.match(/[^0-9]/) === null) {
      this.setState({custom: true, value: s === '' ? '' : parseInt(s)});
    }
  }

  render() {
    const state = this.state;
    return <div className="amount-component" style={{display: 'flex', 'flex-wrap': 'wrap'}}>
      <input type="hidden" value={state.value ? state.value : ''} name={this.hiddenField.name}/>
      {AMOUNTS.map(value => (
        <button key={value} type="button" onClick={() => this.setState({value, custom: false})}
                className={['btn', state.custom || state.value !== value ? 'btn-default' : 'btn-primary'].join(' ')}>
          {value}&nbsp;€
        </button>))}
      <InputGroup>
        <input type="text" className="form-control" placeholder="autre montant" step={1}
                     onChange={e => this.updateWithCustomValue(e.target.value)}
                     value={this.state.custom ? this.state.value.toString() : ''}/>
        <InputGroup.Addon>€</InputGroup.Addon>
      </InputGroup>
      <p>
        {state.value ?
          <strong>
            Après réduction d&apos;impôt, ma contribution nette sera de
            seulement <em className="text-danger">{displayNumber(state.value * 0.34)}</em>&nbsp;€&nbsp;!
          </strong> :
          <strong>En donnant, je profite d&apos;une réduction d&apos;impôt de 66&nbsp;% de la somme donnée !</strong>}
      </p>
    </div>;
  }
}

AmountWidget.propTypes = {
  hiddenField: PropTypes.object
};

export default hot(module)(AmountWidget);
