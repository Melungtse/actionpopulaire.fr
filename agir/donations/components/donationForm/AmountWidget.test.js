import React from "react";
import renderer from "react-test-renderer";

import AmountWidget from "./AmountWidget";

test("AmountWidget looks the right way", () => {
  let currentValue = null,
    component;

  component = renderer.create(
    <AmountWidget
      amount={null}
      onAmountChange={amount => (currentValue = amount)}
      error={null}
    />
  );

  expect(component.toJSON()).toMatchSnapshot();

  // let's click on the 20 â‚¬ button
  const button50 = component.root.find(
    c => c.props.children && c.props.children[0] === 50
  );

  button50.props.onClick();

  expect(currentValue).toEqual(50);

  component = renderer.create(
    <AmountWidget
      amount={50}
      onAmountChange={amount => (currentValue = amount)}
      error={null}
    />
  );

  expect(component.toJSON()).toMatchSnapshot();
});
