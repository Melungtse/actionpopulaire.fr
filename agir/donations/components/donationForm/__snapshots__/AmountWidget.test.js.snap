// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmountWidget looks the right way 1`] = `
<div
  className="amount-component"
>
  <div
    className="form-group"
  >
    <input
      name="amount"
      type="hidden"
      value=""
    />
    <button
      className="btn btn-default sc-bdVaJa hETZJq"
      onClick={[Function]}
      type="button"
    >
      100
       €
    </button>
    <button
      className="btn btn-default sc-bdVaJa hETZJq"
      onClick={[Function]}
      type="button"
    >
      50
       €
    </button>
    <button
      className="btn btn-default sc-bdVaJa hETZJq"
      onClick={[Function]}
      type="button"
    >
      25
       €
    </button>
    <button
      className="btn btn-default sc-bdVaJa hETZJq"
      onClick={[Function]}
      type="button"
    >
      15
       €
    </button>
    <button
      className="btn btn-default sc-bdVaJa hETZJq"
      onClick={[Function]}
      type="button"
    >
      10
       €
    </button>
    <div
      className="input-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="autre montant"
        step={1}
        type="text"
        value=""
      />
      <div
        className="input-group-addon"
      >
        €
      </div>
    </div>
  </div>
  <p>
    <em>
      Si je paye des impôts, je profite d'une réduction d'impôt de 
      <strong>
        66 %
      </strong>
       de la somme donnée !
    </em>
  </p>
</div>
`;

exports[`AmountWidget looks the right way 2`] = `
<div
  className="amount-component"
>
  <div
    className="form-group"
  >
    <input
      name="amount"
      type="hidden"
      value={50}
    />
    <button
      className="btn btn-default sc-bdVaJa hETZJq"
      onClick={[Function]}
      type="button"
    >
      100
       €
    </button>
    <button
      className="btn btn-primary sc-bdVaJa hETZJq"
      onClick={[Function]}
      type="button"
    >
      50
       €
    </button>
    <button
      className="btn btn-default sc-bdVaJa hETZJq"
      onClick={[Function]}
      type="button"
    >
      25
       €
    </button>
    <button
      className="btn btn-default sc-bdVaJa hETZJq"
      onClick={[Function]}
      type="button"
    >
      15
       €
    </button>
    <button
      className="btn btn-default sc-bdVaJa hETZJq"
      onClick={[Function]}
      type="button"
    >
      10
       €
    </button>
    <div
      className="input-group"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="autre montant"
        step={1}
        type="text"
        value=""
      />
      <div
        className="input-group-addon"
      >
        €
      </div>
    </div>
  </div>
  <p>
    <em>
      Si je paye des impôts, après réduction, ma contribution nette sera de seulement
       
      <strong
        className="text-danger"
      >
        17,00 €
      </strong>
       !
    </em>
  </p>
</div>
`;
